test/common/http/conn_manager_impl_test.cc

```diff
@@ -5070,9 +5070,8 @@ TEST_F(HttpConnectionManagerImplTest, FilterContinueAndEndStreamHeaders) {
   Buffer::OwnedImpl fake_input("1234");
   conn_manager_->onData(fake_input, true);
 
-  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, true))
+  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, false))
       .WillOnce(Return(FilterHeadersStatus::ContinueAndEndStream));
-  EXPECT_CALL(*encoder_filters_[1], encodeComplete());
```

Comment:
After fixing the test I realized that this EXPECT fails. After looking at the code, specifically `FilterManager::encodeHeaders`, it turns out the current behavior is that if a filter returns `ContinueAndEndStream` from `encodeHeaders` (to switch `end_stream` to true), `encodeComplete` is not called for that filter.

This behavior also seems like a bug, but I am not entirely sure. To correct we'll have to edit this condition in `FilterManager::encodeHeaders`:
```
if ((*entry)->end_stream_) {
      (*entry)->handle_->encodeComplete();
}
```
to `if ((*entry)->end_stream_ || state_.encoding_headers_only_)` or update `(*entry)->end_stream_` when a filter changes `end_stream`, either through `ContinueAndEndStream` or `ContinueAndDontEndStream`.

Reply:
Yes it's definitely a bug. Nice catch. Can you fix this please? We should definitely be calling encodeComplete() once in all cases when encoding is ended.

Reply:
Should I do it in this PR or a separate PR?